### PR TITLE
Implement TTS and Other REST Refactor

### PR DESCRIPTION
**Pull request by dvonthenen**
Addresses issue: https://github.com/deepgram/deepgram-dotnet-sdk/issues/244
Partially addresses: https://github.com/deepgram/deepgram-dotnet-sdk/issues/234

Major changes include:
- Implements TTS with an Example
- All REST functions now include the ability for custom QueryParameters

There are `TODO`s for logging (and comments) which I will get back to when I have a logging mechanism with verbosity levels worked. Existing issues for Comments https://github.com/deepgram/deepgram-dotnet-sdk/issues/210 and Logging https://github.com/deepgram/deepgram-dotnet-sdk/issues/219 already exist as a reminder.

## Summary by CodeRabbit

- **New Features**
	- Introduced the "Speak" project, enhancing the solution with audio transcription and text-to-speech capabilities.
	- Added support for optional `addons` parameters across various client methods, enabling extended functionality in speech processing tasks.
	- Provided examples demonstrating interaction with the Deepgram API for speech processing.

- **Bug Fixes**
	- Improved code quality by removing unnecessary whitespace and line breaks.

- **Refactor**
	- Updated method signatures to enhance flexibility and maintainability.
	- Streamlined query parameter handling with the introduction of new utility methods.

- **Documentation**
	- Updated descriptions and added documentation for new classes and methods related to the Speak API.

### CODE REVIEW

File: Deepgram/Abstractions/AbstractRestClient.cs [hunk 1]
```diff
@@ -2,8 +2,11 @@
 // Use of this source code is governed by a MIT license that can be found in the LICENSE file.
 // SPDX-License-Identifier: MIT
 
+using System.Linq.Expressions;
+using System.Net.Http;
 using Deepgram.Encapsulations;
 using Deepgram.Models.Authenticate.v1;
+using Deepgram.Utilities;
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Abstractions/AbstractRestClient.cs:9]: Fixing!

File: Deepgram/Abstractions/AbstractRestClient.cs [hunk 2]
```diff
@@ -68,11 +64,76 @@
...
+            await response.Content.CopyToAsync(stream);
+
+            return (result, stream);
+        }
+        catch (Exception ex)
+        {
+            Log.Exception(_logger, "POST", ex);
+            throw;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Abstractions/AbstractRestClient.cs:123]: Yup, I noticed that as well!

File: Deepgram/Models/Speak/v1/TextSource.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Deepgram.Models.Speak.v1;
+
+public class TextSource(string text)
+{
+    /// <summary>
+    /// Text of the words to speak
+    /// </summary>
+    [JsonPropertyName("text")]
+    public string? Text { get; set; } = text;
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Models/Speak/v1/TextSource.cs:13]: This should be ok. Dismissing this.

File: examples/speak/Program.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+using Deepgram.Models.Speak.v1;
+using System.Text.Json;
+
+namespace SampleApp
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            var apiKey = "REPLACE-WITH-YOUR-API-KEY";
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/speak/Program.cs:11]: Sure why not!

File: Deepgram/Models/Speak/v1/SyncResponse.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
+﻿// Copyright 2021-2024 Deepgram .NET SDK contributors. All Rights Reserved.
+// Use of this source code is governed by a MIT license that can be found in the LICENSE file.
+// SPDX-License-Identifier: MIT
+
+namespace Deepgram.Models.Speak.v1;
+
+public record SyncResponse
+{
+    /// <summary>
+    /// TODO
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Models/Speak/v1/SyncResponse.cs:10]: Will take care of Logging and Comment TODOs later

File: Deepgram/Utilities/QueryParameterUtil.cs [hunk 6]
```diff
@@ -3,23 +3,47 @@
...
+        var query = HttpUtility.ParseQueryString(uriBuilder.Query);
+
+        for (var i = 0; i < parameters.Count; i++)
+        {
+            query[parameters.ElementAt(i).Key] = parameters.ElementAt(i).Value;
+        }
+
+        uriBuilder.Query = query.ToString();
+
+        return uriBuilder.ToString();
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Utilities/QueryParameterUtil.cs:43]: Sure!

File: Deepgram/Clients/Speak/v1/Client.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    public async Task<AsyncResponse> StreamCallBack(TextSource source, string? callBack, SpeakSchema? speakSchema, CancellationToken cancellationToken = default, Dictionary<string, string>? addons = null)
+    {
+        VerifyOneCallBackSet(nameof(StreamCallBack), callBack, speakSchema);
+        if (callBack != null)
+            speakSchema.CallBack = callBack;
+
+        var stringedOptions = QueryParameterUtil.GetParameters(speakSchema, addons);
+        return await PostAsync<AsyncResponse>(
+            $"{UriSegments.SPEAK}?{stringedOptions}",
+            RequestContentUtil.CreatePayload(source), cancellationToken);
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Speak/v1/Client.cs:123]: I think this is easier to read!

File: Deepgram/Models/Speak/v1/SpeakSchema.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /// <summary>
+    /// AI model used to process submitted audio
+    /// <see href="https://developers.deepgram.com/docs/model">
+    /// </summary>
+    [JsonPropertyName("model")]
+    public string? Model { get; set; }
+
+    /// <summary>
+    /// TODO
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Models/Speak/v1/SpeakSchema.cs:17]: Will address TODOs for logging and comments later

**Review by SandraRodgers — APPROVED** [review]

**Pull request merged**